### PR TITLE
retry on receiving 5xx

### PR DESCRIPTION
**Pull request by smusali**
Retry should happen not only when it gets `http` and/or network-related exceptions but also when it receives `5xx` status code from the server which is meaning `Server Error`

### CODE REVIEW

**Comment by weizou19** [comment]
LGTM except the test failing

File: lib/linebuffer.js [hunk 1]
```diff
@@ -24,6 +24,31 @@ var dccount = 0;
 var httperr;
 var queue;
 
+// check response to see need for buffering into queue
+const checkResponse = (error, response, body) => {
```
Comment (dchai76) [inline lib/linebuffer.js]: I would name this something like `needsRetry` - that it's returning a bool, and what that value is meant to indicate is not clear from the name.

File: lib/linebuffer.js [hunk 2]
```diff
@@ -24,6 +24,31 @@ var dccount = 0;
...
+    if (error) {
+        debug('req error ' + error.code + ': ' + error);
+        if (!httperr) { httperr = Date.now(); }
```
Comment (dchai76) [inline lib/linebuffer.js]: This precedes you so you don't have to change it but as a reader, this is pretty mystifying on a couple levels - what the var means and why it's being set to a date.

File: lib/get-distro.js [hunk 3]
```diff
@@ -30,7 +30,7 @@ module.exports = (path, callback) => {
     if (os.platform() === 'win32') {
         return getDist(callback);
     }
-    
+
```
[the remark this answers is not in the document]
Comment (smusali) [reply lib/get-distro.js]: it happened due to `eslint --fix`

[the remark this answers is not in the document]
Comment (dchai76) [reply lib/get-distro.js]: Right, I recommend you only run `eslint` on modified files

File: lib/linebuffer.js [hunk 4]
```diff
@@ -24,6 +24,31 @@ var dccount = 0;
 var httperr;
 var queue;
 
+// check response to see need for buffering into queue
+const checkResponse = (error, response, body) => {
+    if (error) {
+        debug('req error ' + error.code + ': ' + error);
```
[the remark this answers is not in the document]
Comment (smusali) [reply lib/linebuffer.js]: Done! Couldn't get `error` part - can you clarify, @dchai76?

File: lib/linebuffer.js (hunk 1 above)
Comment (smusali) [reply lib/linebuffer.js]: Good naming - liked - done!

File: lib/linebuffer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (smusali) [reply lib/linebuffer.js]: Done!

File: lib/linebuffer.js (hunk 2 above)
Comment (smusali) [reply lib/linebuffer.js]: renaming makes it easier to understand why it's set to datetime object

**Comment by smusali** [comment]
@dchai76 @weizou19 can you do second round of review?

File: lib/linebuffer.js [hunk 5]
```diff
@@ -21,18 +20,43 @@ var bufferStats = {
 };
 var config = require('./config');
 var dccount = 0;
-var httperr;
+var http_error_time;
```
Comment (dchai76) [inline lib/linebuffer.js]: `httpErrorTime`

File: lib/linebuffer.js [hunk 6]
```diff
@@ -48,75 +72,56 @@ const flush = () => {
...
-            } else if (res.statusCode !== 200) {
-                debug('req failed: ' + body);
-                if (bufferStats.flushcount % 10 === 0) { utils.log('Request failed: ' + body); }
-
-            } else {
-                if (httperr) {
-                    debug('req succeeded, re-enabling');
-                    httperr = null;
-                }
+                return; // prevent fallthru
```
Comment (dchai76) [inline lib/linebuffer.js]: `fallthrough`

Comment (smusali) [reply lib/linebuffer.js]: done!

File: lib/linebuffer.js (hunk 5 above)
Comment (smusali) [reply lib/linebuffer.js]: done!

File: lib/linebuffer.js [hunk 7]
```diff
@@ -21,18 +20,43 @@ var bufferStats = {
...
+            return response.statusCode >= 500;
+        } else {
+            if (httpErrorTime) {
+                debug('req succeeded, re-enabling');
+                httpErrorTime = null;
+            }
+        }
+    }
+
+    return false;
```
Comment (weizou19) [inline lib/linebuffer.js]: I'd prefer to return as early as possible, more readable, and less error-prone:
```
if (!response) {
    return false;
}

if (response.statusCode !== 200) {
    debug(`req failed: ${body}`);
    if (bufferStats.flushcount % 10 === 0) {
        utils.log(`request failed: ${body}`);
    }
    return response.statusCode >= 500;
}

if (httpErrorTime) {
    debug('req succeeded, re-enabling');
    httpErrorTime = null;
}
return false;
```

File: lib/linebuffer.js [hunk 8]
```diff
@@ -21,18 +20,43 @@ var bufferStats = {
...
-        if (Date.now() - httperr < config.HTTP_TIMEOUT / 2) {
+    if (httpErrorTime && config.TRANSPORT === 'http') {
+        if (Date.now() - httpErrorTime < config.HTTP_TIMEOUT / 2) {
             dccount = buf.length;
             return;
         } else {
-            httperr = Date.now(); // reset
+            httpErrorTime = Date.now(); // reset
             debug('req re-attempting');
         }
```
Comment (weizou19) [inline lib/linebuffer.js]: No need for the above `else` block, as the `if` block already `return`

**Comment by weizou19** [comment]
Also, do we have test case to cover the new logic?

File: lib/linebuffer.js (hunk 7 above)
Comment (smusali) [reply lib/linebuffer.js]: This part is done but I don't know why it doesn't show as `Outdated` - can you check this part again, @weizou19?

File: lib/linebuffer.js (hunk 8 above)
Comment (smusali) [reply lib/linebuffer.js]: Done!

File: lib/linebuffer.js (hunk 7 above)
Comment (smusali) [reply lib/linebuffer.js]: Done!

**Pull request merged**